### DISCUSSION TITLE
Can't make OIDC to work

### DISCUSSION THREAD

**Discussion by aleks5323** [outline/outline#2606] [Self Hosting] [answered]
Hi there!
I wonder if someone can help me to point out what am i misunderstanding and doing wrong. I have outline Outline setup and running in Docker container, as well as other needed services. All the services are behind nginx proxy manager and have self signed certificates. I have successfully (as far as i can tell by the logs) connected KeyCloak IDP to outline but have one problem - when i hit log in button and enter my keycloak credentials i get redirected back to outline with an error "Authentication failed – we were unable to sign you in at this time. Please try again." and log says either <br> `InternalOAuthError: Failed to obtain access token`
`...` 
`code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE'`
or
 `InternalOAuthError: Failed to obtain access token`
`...` 
`code: 'DEPTH_ZERO_SELF_SIGNED_CERT'`
As far as I understand all this is because i host it locally and because of self sign cert (i have no plans to make this accessible for public net).
So the question is - Is there a workaround to make it work locally without making it work through a public domain and getting real ssl? Sorry if this is a really silly question to ask, but i'm new to this kind of things.

P.S. Sorry for bad grammar, I hope I have explained everything clearly.

**Comment by tommoor** [comment]
I've not seen this error before but it does seem related to your self signed cert, there's no expectation or testing of Outline working in this environment so you're on your own I'm afraid. 

If you want something to make having it on the public internet easier I've had good success with this docker image as an automatic https proxy: https://github.com/SteveLTN/https-portal

**Comment by matteo-bruni** [comment, marked as answer]
I'm using outline with authentik (instead of keycloack) and every service is local including the certs for each service (I'm using smallstep CA to generate them). If your authentication provider has a self-signed certificate it will not connect. Usually, there are env variables to disable check on a particular connection but passport-oauth does not seem to have one and you can't add certs to the outline container (unless you rebuild the image)

I solved it by adding this env variable to outline NODE_TLS_REJECT_UNAUTHORIZED=0 to disable the check on node. 
BEWARE that it will disable the check for EVERY connection (i don't mind since I have all services available only from lan)

**Reply by aleks5323** [reply]
That's exactly what i did before i saw your reply, but thanks anyway)
And i think i found another way of solving this problem. Instead of using NODE_TLS_REJECT_UNAUTHORIZED=0, i put NODE_EXTRA_CA_CERTS='./certs/ca.pem' in 'start' script inside package.json file and i guess it works fine. Not sure if that's a right way of setting variable. Correct me if i'm wrong.

**Reply by tommoor** [reply]
Great stuff!

**Reply by matmair** [reply]
@tommoor this just saved me from creating my own image. Maybe this could be added to the [OIDC docs](https://wiki.generaloutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/oidc-8CPBm6uC0I#h-examples)? I can imagine ppl. using their own IDPs run into this problem often.

**Reply by tommoor** [reply]
👍 I added a note to the "Troubleshooting" section

**Reply by matmair** [reply]
Thank you. Great application!

**Reply by axelsimon** [reply]
> Instead of using NODE_TLS_REJECT_UNAUTHORIZED=0, i put NODE_EXTRA_CA_CERTS='./certs/ca.pem' in 'start' script inside package.json file and i guess it works fine. Not sure if that's a right way of setting variable. Correct me if i'm wrong.

For anyone else reading this, you'll want to put `NODE_EXTRA_CA_CERTS=./certs/ca.pem` (no quotes) as an environment variable. Otherwise, node will look for the path with the single quotes and fail to find the file!
